heft-plugins/heft-webpack4-plugin/src/Webpack4Plugin.ts

```diff
@@ -148,130 +183,255 @@ export default class Webpack4Plugin implements IHeftTaskPlugin<IWebpack4PluginOp
-          )
-        );
-      }
+      // Determine how we will run the compiler. When serving, we will run the compiler
+      // via the webpack-dev-server. Otherwise, we will run the compiler directly.
+      if (this._serve) {
+        const defaultDevServerOptions: TWebpackDevServer.Configuration = {
+          host: 'localhost',
+          devMiddleware: {
+            publicPath: '/',
+            stats: {
+              cached: false,
+              cachedAssets: false,
+              colors: heftConfiguration.terminalProvider.supportsColor
+            }
+          },
+          client: {
+            logging: 'info'
+          },
+          port: 8080,
+          onListening: (server: TWebpackDevServer) => {
+            const addressInfo: AddressInfo | string | undefined = server.server?.address() as AddressInfo;
+            if (addressInfo) {
+              const address: string =
+                typeof addressInfo === 'string' ? addressInfo : `${addressInfo.address}:${addressInfo.port}`;
+              taskSession.logger.terminal.writeLine(`Started Webpack Dev Server at https://${address}`);
+            }
+          }
+        };
 
-      let stats: TWebpack.Stats | TWebpack.compilation.MultiStats | undefined;
-      if (watchMode) {
-        try {
-          stats = await LegacyAdapters.convertCallbackToPromise(
-            (compiler as TWebpack.Compiler).watch.bind(compiler),
-            {}
-          );
-        } catch (e) {
-          logger.emitError(e as Error);
+        // Obtain the devServerOptions from the webpack configuration, and combine with the default options
+        let devServerOptions: TWebpackDevServer.Configuration;
+        if (Array.isArray(webpackConfiguration)) {
+          const filteredDevServerOptions: TWebpackDevServer.Configuration[] = webpackConfiguration
+            .map((configuration) => configuration.devServer)
+            .filter((devServer): devServer is TWebpackDevServer.Configuration => !!devServer);
+          if (filteredDevServerOptions.length > 1) {
+            taskSession.logger.emitWarning(
+              new Error(`Detected multiple webpack devServer configurations, using the first one.`)
+            );
+          }
+          devServerOptions = { ...defaultDevServerOptions, ...filteredDevServerOptions[0] };
+        } else {
+          devServerOptions = { ...defaultDevServerOptions, ...webpackConfiguration.devServer };
         }
+
+        // Add the certificate and key to the devServerOptions
+        const certificateManager: CertificateManager = new CertificateManager();
+        const certificate: ICertificate = await certificateManager.ensureCertificateAsync(
+          true,
+          taskSession.logger.terminal
+        );
+        devServerOptions = {
+          ...devServerOptions,
+          server: {
+            type: 'https',
+            options: {
+              key: certificate.pemKey,
+              cert: certificate.pemCertificate
+            }
```

Comment:
Should we blindly overwrite the `server` and `server.options` properties?

Reply:
I can check to see if `server` has a preset and avoid populating these values (and loading the cert) if not necessary.